src/serde/reader.rs

```diff
@@ -54,15 +58,39 @@ impl<R: Read, E: ByteOrder> Deserializer<R, E> {
         self.read_bytes(size_of::<T>() as u64)
     }
 
+    fn read_vec_small(&mut self, len: usize) -> Result<Vec<u8>> {
+        let mut result = Vec::with_capacity(len);
+        unsafe { result.set_len(len); }
+        try!(self.reader.read_exact(&mut result));
+        Ok(result)
+    }
+
+    fn read_vec_large(&mut self, mut len: usize, block: usize) -> Result<Vec<u8>> {
```

Comment:
Oh wait, I think it's worse than that actually.  I think that it should read `Vec::new()`, because the very next thing that gets executed is a call to `reserve`.  

In my implementation I think I'm always over-allocating.